File: programmers/kakao#2/python/KAKAO25.py
from sys import setrecursionlimit
setrecursionlimit(10 ** 9)
class Node:
    def __init__(self, value, x, y, left = None, right = None):
        self.value = value
        self.left = left
        self.right = right
        self.x = x
        self.y = y
    
    def __str__(self):
        return f'value = {self.value}, x = {self.x}, y = {self.y}'

index = 0
def solution(nodeinfo):
    global index
    nodes = [Node(index, info[0], info[1]) for index, info in enumerate(nodeinfo, 1)]    
    nodes.sort(key=lambda n: [-n.y, n.x])
    
    root = nodes[0]
    for i in range(1, len(nodes)):
        addNode(root, nodes[i])
    
    answer = [[0 for _ in range(len(nodes))] for _ in range(2)]
    preorder(answer[0], root)
    index = 0
    postorder(answer[1], root)
    return answer

def preorder(seq, node):
    global index
    if node is not None:
        seq[index] = node.value
        index += 1
        preorder(seq, node.left)
        preorder(seq, node.right)


def postorder(seq, node):
    global index
    if node is not None:
        postorder(seq, node.left)
        postorder(seq, node.right)
        seq[index] = node.value
        index += 1

def addNode(parent, child):
    if parent.x > child.x:
        if parent.left is None:
            parent.left = child
            return
        addNode(parent.left, child)
        return
    else:
        if parent.right is None:
            parent.right = child
            return
        addNode(parent.right, child)
        return

'''
[카카오 2019 공채] 길 찾기 게임
해결: X
'''

File: 0_4.programmers/kakao#2/java/KAKAO25.java
import java.util.*;
class KAKAO25 {
    public static class Node {
        int value;
        int x, y;
        Node left, right;
        
        Node(int value, int x, int y) {
            this.value = value;
            this.x = x;
            this.y = y;
        }
        
        public String toString() {
            return "value = " + value;
        }
    }
    private int index = 0;
    public int[][] solution(int[][] nodeinfo) {
        List<Node> nodes = new ArrayList<>();

        for (int i=0;i<nodeinfo.length;i++) {
            nodes.add(new Node(i+1, nodeinfo[i][0], nodeinfo[i][1]));
        }
        
        sortByCoordinate(nodes);
        buildTree(nodes);
        
        int[][] answer = new int[2][nodes.size()];
        preorder(answer[0], nodes.get(0));
        index = 0;
        postorder(answer[1], nodes.get(0));
        return answer;
    }
    
    private void postorder(int[] seq, Node node) {
        if (node != null) {
            postorder(seq, node.left);
            postorder(seq, node.right);
            seq[index++] = node.value;
        }
    }
    
    private void preorder(int[] seq, Node node) {
        if (node != null) {
            seq[index++] = node.value;
            preorder(seq, node.left);
            preorder(seq, node.right);
        }
    }
    
    private void buildTree(List<Node> nodes) {
        Node root = nodes.get(0);
        
        for (int i=1;i<nodes.size();i++) {
            addNode(root, nodes.get(i));
        }
    }
    
    private void addNode(Node parent, Node child) {
        if (parent.x > child.x) {
            if (parent.left == null) {
                parent.left = child;
            } else {
                addNode(parent.left, child);
            }
        } else {
            if (parent.right == null) {
                parent.right = child;
            } else {
                addNode(parent.right, child);
            }
        }
    }
    
    private void sortByCoordinate(List<Node> nodes) {
        Collections.sort(nodes, new Comparator<>() {
           public int compare(Node n1, Node n2) {
               if (n1.y != n2.y) 
                   return n2.y - n1.y;
               return n1.x - n2.x;
           } 
        });
    }
}
/***
 * [카카오 2019 공채] 길 찾기 게임
 * 해결: X
 */